### PR TITLE
Cleanup UPDGRADE.md for the 4.0 release

### PR DESCRIPTION
**Pull request by morozov**

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -624,60 +606,14 @@ The `ServerInfoAwareConnection::requiresQueryForServerVersion()` method has been
...
-- Class `Doctrine\DBAL\Cache\ResultCacheStatement` was made final.
-- Class `Doctrine\DBAL\Cache\ArrayStatement` was made final.
-- Class `Doctrine\DBAL\Schema\Synchronizer\SingleDatabaseSynchronizer` was made final.
-- Class `Doctrine\DBAL\Portability\Statement` was made final.
-
 ## BC BREAK: Changes in the `Doctrine\DBAL\Schema` API
 
 - Column precision no longer defaults to 10. The default value is NULL.
 - Asset names are no longer nullable. An empty asset name should be represented as an empty string.
-- `Doctrine\DBAL\Schema\AbstractSchemaManager::_getPortableTriggersList()` and `::_getPortableTriggerDefinition()` have been removed.
```
Comment (morozov) [inline UPGRADE.md:665]: See https://github.com/doctrine/dbal/pull/4752 (3.2.0).

File: UPGRADE.md [hunk 2]
```diff
@@ -798,27 +694,10 @@ Similarly to `PDOStatement::fetchColumn()`, DBAL statements throw an exception i
 
 Similarly to the drivers based on `pdo_pgsql` and `pdo_sqlsrv`, `OCI8Statement::execute()` and `MySQLiStatement::execute()` do not longer ignore redundant parameters.
 
-## BC BREAK: `Doctrine\DBAL\Types\Type::getDefaultLength()` removed
-
-The `Doctrine\DBAL\Types\Type::getDefaultLength()` method has been removed as it served no purpose.
```
Comment (morozov) [inline UPGRADE.md:803]: See https://github.com/doctrine/dbal/pull/4278 (3.0.0).

File: UPGRADE.md [hunk 3]
```diff
@@ -798,27 +694,10 @@ Similarly to `PDOStatement::fetchColumn()`, DBAL statements throw an exception i
...
-
-## BC BREAK: `Doctrine\DBAL\Types\Type::__toString()` removed
-
-Relying on string representation was discouraged and has been removed.
```
Comment (morozov) [inline UPGRADE.md:807]: See https://github.com/doctrine/dbal/pull/4082 (3.0.0).

File: UPGRADE.md [hunk 4]
```diff
@@ -798,27 +694,10 @@ Similarly to `PDOStatement::fetchColumn()`, DBAL statements throw an exception i
...
-
 ## BC BREAK: The `NULL` value of `$offset` in LIMIT queries is not allowed
 
 The `NULL` value of the `$offset` argument in `AbstractPlatform::(do)?ModifyLimitQuery()` methods is no longer allowed. The absence of the offset should be indicated with a `0` which is now the default value.
 
-## BC BREAK: Removed support for DB-generated UUIDs
-
-The support for DB-generated UUIDs was removed as non-portable.
```
Comment (morozov) [inline UPGRADE.md:815]: See https://github.com/doctrine/dbal/pull/4082 (3.0.0).

File: UPGRADE.md [hunk 5]
```diff
@@ -798,27 +694,10 @@ Similarly to `PDOStatement::fetchColumn()`, DBAL statements throw an exception i
...
-Please generate UUIDs on the application side (e.g. using [ramsey/uuid](https://packagist.org/packages/ramsey/uuid)).
-
-## BC BREAK: Removed Doctrine\DBAL\Version
-
-The Doctrine\DBAL\Version class is no longer available: please refrain from checking the DBAL version at runtime.
```
Comment (morozov) [inline UPGRADE.md:820]: See https://github.com/doctrine/dbal/pull/3712 (3.0.0).

File: UPGRADE.md [hunk 6]
```diff
@@ -687,48 +623,12 @@ Table columns are no longer indexed by column name. Use the `name` attribute of
...
-`Connection::ping()` and `PingableConnection::ping()` no longer return a boolean value. They will throw an exception in case of failure.
 
 ## BC BREAK PostgreSqlPlatform ForeignKeyConstraint support for `feferred` misspelling removed
 
 `PostgreSqlPlatform::getAdvancedForeignKeyOptionsSQL()` had a typo in it in 2.x. Both the option name
 `feferred` and `deferred` were supported in `2.x` but the misspelling was removed in 3.x.
 
-## BC BREAK `AbstractSchemaManager::extractDoctrineTypeFromComment()` changed, `::removeDoctrineTypeFromComment()` removed
-
-`AbstractSchemaManager::extractDoctrineTypeFromComment()` made `protected`. It takes the comment by reference, removes the type annotation from it and returns the extracted Doctrine type.
```
Comment (morozov) [inline UPGRADE.md:730]: See https://github.com/doctrine/dbal/pull/5107 (4.0.0).

File: UPGRADE.md [hunk 7]
```diff
@@ -20,9 +20,9 @@ The following methods have been removed:
 1. `Doctrine\DBAL\Statement::bindParam()`,
 2. `Doctrine\DBAL\Driver\Statement::bindParam()`.
 
-## BC BREAK: made parameter type in driver-level `Statement::bind*()` methods required.
+## BC BREAK: made parameter type in driver-level `Statement::bindValue()` required.
 
-The `$type` parameter of the driver-level `Statement::bindParam()` and `::bindValue()` has been made required.
```
Comment (morozov) [inline UPGRADE.md:25]: See https://github.com/doctrine/dbal/pull/5565 (4.0.0).

File: UPGRADE.md [hunk 8]
```diff
@@ -570,10 +570,6 @@ The following schema- and namespace-related methods have been removed:
 - `AbstractSchemaManager::getPortableNamespaceDefinition()`,
 - `PostgreSQLSchemaManager::getSchemaNames()`.
 
-## Removed the `$driverOptions` argument of `PDO\Statement::bindParam()` and `PDO\SQLSrv\Statement::bindParam()`
-
-The `$driverOptions` argument of `PDO\Statement::bindParam()` and `PDO\SQLSrv\Statement::bindParam()` has been removed. The specifics of binding a parameter to the statement should be specified using the `$type` argument.
```
Comment (morozov) [inline UPGRADE.md:575]: See https://github.com/doctrine/dbal/pull/5565 (4.0.0).

File: UPGRADE.md [hunk 9]
```diff
@@ -624,60 +606,14 @@ The `ServerInfoAwareConnection::requiresQueryForServerVersion()` method has been
 
 Table columns are no longer indexed by column name. Use the `name` attribute of the column instead.
 
-## BC BREAK: Classes made final
```
Comment (morozov) [inline UPGRADE.md:627]: All these classes are already final since 3.0.0 or no longer exist.

File: UPGRADE.md [hunk 10]
```diff
@@ -624,60 +606,14 @@ The `ServerInfoAwareConnection::requiresQueryForServerVersion()` method has been
...
 
 ## BC BREAK: Changes in the `Doctrine\DBAL\Event` API
 
 - `SchemaAlterTableAddColumnEventArgs::addSql()` and the same method in other `SchemaEventArgs`-based classes no longer accept an array of SQL statements. They accept a variadic string.
-- `ConnectionEventArgs::getDriver()`, `::getDatabasePlatform()` and `::getSchemaManager()` methods have been removed. The connection information can be obtained from the connection which is available via `::getConnection()`.
-- `SchemaColumnDefinitionEventArgs::getDatabasePlatform()` and `SchemaIndexDefinitionEventArgs::getDatabasePlatform()` have been removed for the same reason as above.
-
-## BC BREAK: Changes in the `Doctrine\DBAL\Connection` API
-
-- The following methods have been removed as leaking internal implementation details: `::getHost()`, `::getPort()`, `::getUsername()`, `::getPassword()`.
```
Comment (morozov) [inline UPGRADE.md:675]: See https://github.com/doctrine/dbal/pull/4082 (3.0.0).

File: UPGRADE.md [hunk 11]
```diff
@@ -624,60 +606,14 @@ The `ServerInfoAwareConnection::requiresQueryForServerVersion()` method has been
...
-- The `::getDatabase()` method can now return null which means that no database is currently selected.
-
-## BC BREAK: Changes in `Doctrine\DBAL\Driver\SQLSrv\LastInsertId`
-
-- The class stores the last inserted ID as a nullable string, not an integer, which is reflected in the method signatures.
```
Comment (morozov) [inline UPGRADE.md:680]: See https://github.com/doctrine/dbal/pull/4514 (3.1.0).

File: UPGRADE.md [hunk 12]
```diff
@@ -687,48 +623,12 @@ Table columns are no longer indexed by column name. Use the `name` attribute of
 - Property `Doctrine\DBAL\Schema\TableDiff::$newName` is now optionally null instead of false.
 - Method `Doctrine\DBAL\Schema\AbstractSchemaManager::tablesExist()` no longer accepts a string. Use `Doctrine\DBAL\Schema\AbstractSchemaManager::tableExists()` instead.
 - Method `Doctrine\DBAL\Schema\OracleSchemaManager::createDatabase()` no longer accepts `null` for `$database` argument.
-- Removed unused method `Doctrine\DBAL\Schema\AbstractSchemaManager::_getPortableFunctionsList()`
-- Removed unused method `Doctrine\DBAL\Schema\AbstractSchemaManager::_getPortableFunctionDefinition()`
-- Removed unused method `Doctrine\DBAL\Schema\OracleSchemaManager::_getPortableFunctionDefinition()`
-- Removed unused method `Doctrine\DBAL\Schema\SqliteSchemaManager::_getPortableTableIndexDefinition()`
```
Comment (morozov) [inline UPGRADE.md:693]: See https://github.com/doctrine/dbal/pull/4082 (3.0.0).

File: UPGRADE.md [hunk 13]
```diff
@@ -687,48 +623,12 @@ Table columns are no longer indexed by column name. Use the `name` attribute of
...
- * Removed `AbstractPlatform::DATE_INTERVAL_UNIT_*` and `AbstractPlatform::TRIM_*` constants.
- * Removed `MysqlSessionInit` listener.
- * Removed `MysqlPlatform::getCollationFieldDeclaration()`.
- * Removed `AbstractPlatform::getIdentityColumnNullInsertSQL()`.
- * Removed `Table::addUnnamedForeignKeyConstraint()` and `Table::addNamedForeignKeyConstraint()`.
- * Removed `Table::renameColumn()`.
- * Removed `SQLParserUtils::getPlaceholderPositions()`.
- * Removed `AbstractSchemaManager::getFilterSchemaAssetsExpression()`, `Configuration::getFilterSchemaAssetsExpression()`
-   and `Configuration::getFilterSchemaAssetsExpression()`.
- * `SQLParserUtils::*_TOKEN` constants made private.
```
Comment (morozov) [inline UPGRADE.md:717]: See https://github.com/doctrine/dbal/pull/4082 (3.0.0).

File: UPGRADE.md [hunk 14]
```diff
@@ -20,9 +20,9 @@ The following methods have been removed:
 1. `Doctrine\DBAL\Statement::bindParam()`,
 2. `Doctrine\DBAL\Driver\Statement::bindParam()`.
 
-## BC BREAK: made parameter type in driver-level `Statement::bind*()` methods required.
+## BC BREAK: made parameter type in driver-level `Statement::bindValue()` required.
 
-The `$type` parameter of the driver-level `Statement::bindParam()` and `::bindValue()` has been made required.
+The `$type` parameter of the driver-level `Statement::bindValue()` has been made required.
```
Comment (morozov) [inline UPGRADE.md:25]: See https://github.com/doctrine/dbal/pull/5565 (4.0.0).

File: UPGRADE.md (hunk 8 above)
Comment (morozov) [inline UPGRADE.md:575]: See https://github.com/doctrine/dbal/pull/5565 (4.0.0).

**Review by derrabus — APPROVED** [review]

**Pull request merged**